tests/end_to_end/end_to_end.cpp

```diff
@@ -43,6 +46,76 @@ prepare_udp_packet(uint16_t udp_length)
     return packet;
 }
 
+const std::array<uint8_t, 6> _test_source_mac = {0, 1, 2, 3, 4, 5};
+const std::array<uint8_t, 6> _test_destination_mac = {0xa, 0xb, 0xc, 0xd, 0xe, 0xf};
+const in_addr _test_source_ip = {10, 0, 0, 1};
+const in_addr _test_destination_ip = {20, 0, 0, 1};
+
+typedef class _udp_packet
+{
+  public:
+    _udp_packet(
+        _In_ const std::array<uint8_t, 6>& source_mac = _test_source_mac,
+        _In_ const std::array<uint8_t, 6> destination_mac = _test_destination_mac,
+        _In_ const in_addr* source_ip = &_test_source_ip,
+        _In_ const in_addr* destination_ip = &_test_destination_ip,
+        uint16_t datagram_length = 1024)
+    {
+        UNREFERENCED_PARAMETER(source_mac);
+        UNREFERENCED_PARAMETER(destination_mac);
+        _packet = prepare_udp_packet(sizeof(ebpf::UDP_HEADER) + datagram_length);
+        set_mac_addresses(source_mac, destination_mac);
+        set_ip_addresses(source_ip, destination_ip);
+    }
+    uint8_t*
+    data()
+    {
+        return _packet.data();
+    }
+    size_t
+    size()
+    {
+        return _packet.size();
+    }
+
+    void
+    set_source_port(uint16_t source_port)
+    {
+        auto eth = reinterpret_cast<ebpf::ETHERNET_HEADER*>(_packet.data());
+        auto ipv4 = reinterpret_cast<ebpf::IPV4_HEADER*>(eth + 1);
+        auto udp = reinterpret_cast<ebpf::UDP_HEADER*>(ipv4 + 1);
+        udp->srcPort = source_port;
+    }
+
+    void
+    set_destination_port(uint16_t destination_port)
+    {
+        auto eth = reinterpret_cast<ebpf::ETHERNET_HEADER*>(_packet.data());
+        auto ipv4 = reinterpret_cast<ebpf::IPV4_HEADER*>(eth + 1);
+        auto udp = reinterpret_cast<ebpf::UDP_HEADER*>(ipv4 + 1);
+        udp->destPort = destination_port;
+    }
+
+  private:
+    std::vector<uint8_t> _packet;
+    void
+    set_mac_addresses(_In_ const std::array<uint8_t, 6>& source_mac, _In_ const std::array<uint8_t, 6>& destination_mac)
+    {
+        auto eth = reinterpret_cast<ebpf::ETHERNET_HEADER*>(_packet.data());
+        memcpy(eth->Source, source_mac.data(), sizeof(eth->Source));
+        memcpy(eth->Destination, destination_mac.data(), sizeof(eth->Destination));
+    }
+    void
+    set_ip_addresses(_In_ const in_addr* source_address, _In_ const in_addr* destination_address)
```

Comment:
bad name, since "ip addresses" is version agnostic.  So either make the args be version agnostic, or else there should be separate set_ipv4_addresses and set_ipv6_addresses functions.